### PR TITLE
[TTI][TLI] Support scalable immediates with isLegalAddImmediate

### PR DESCRIPTION
**Pull request by huntergr-arm**
Adds a second parameter (default to 0) to isLegalAddImmediate, to represent a scalable immediate.

Extends the AArch64 implementation to match immediates based on vscale * base_register_size * immediate in the range [-32,31].

See the vscale-aware LSR RFC for reference: https://discourse.llvm.org/t/rfc-vscale-aware-loopstrengthreduce/77131

### CODE REVIEW

**Comment by davemgreen** [comment]
Can you remind us where these functions are used? It might be simpler to have separate methods and move the scalable vs normal immediate check out a level.

**Review by RKSimon — COMMENTED** [review]
I'm confused as to how this will work - how do you indicate when to test for a legal scalar or scalable immediate? Suppose you want to know if 0 is legal?

File: llvm/include/llvm/Analysis/TargetTransformInfo.h [hunk 1]
```diff
@@ -696,6 +696,12 @@ class TargetTransformInfo {
   /// immediate without having to materialize the immediate into a register.
   bool isLegalAddImmediate(int64_t Imm) const;
 
+  /// Return true if the specified immediate is legal add of a scalable
+  /// immediate, that is the target has add instructions which can add a
+  /// register with the immediate (multiplied by vscale) without having to
+  /// materialize the immediate into a register.
```
Comment (davemgreen) [inline llvm/include/llvm/Analysis/TargetTransformInfo.h:702]: Maybe:
"Return true if the add of the specified scalable immediate is legal, that is..."

File: llvm/unittests/Target/AArch64/Immediates.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+const std::initializer_list<TestCase> Tests = {
+    // ScalableImm, Result
+    // No change, easily 'supported'
+    {0, true},
+
+    // addvl increments by whole registers, range [-32,31]
+    // +(16 * vscale), one register's worth
+    {16, true},
+    // +(8 * vscale), half a register's worth
+    {8, false},
```
Comment (davemgreen) [inline llvm/unittests/Target/AArch64/Immediates.cpp:29]: 8 could be a `inch` if that wouldn't cause problems with your analysis. Same for 2 and 4.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -16595,6 +16595,21 @@ bool AArch64TargetLowering::isLegalAddImmediate(int64_t Immed) const {
   return IsLegal;
 }
 
+bool AArch64TargetLowering::isLegalAddScalableImmediate(int64_t Imm) const {
+  // Scalable immediates require SVE support.
+  if (!Subtarget->hasSVE())
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:16600]: Should all of this require sve2? Otherwise we don't generate the instructions.

**Review by davemgreen — COMMENTED** [review]
This is looking sensible to me, I have some comments below from experimenting with some values.

**Review by davemgreen — APPROVED** [review]
Thanks. This looks sensible to me.

**Pull request merged**